OpenRA.Game/GameRules/RulesetWatcher.cs

```diff
@@ -0,0 +1,162 @@
+
+namespace OpenRA.GameRules
+{
+	public class RulesetWatcher : IDisposable
+	{
+		readonly World world;
+		readonly ModData modData;
+		readonly Dictionary<string, string> watchFiles = new();
+		readonly Timer timer;
+		readonly HashSet<string> fileQueue = new();
+		readonly FileSystemWatcher watcher;
+
+		volatile bool isEnabled;
+		bool isDisposed;
+
+		public bool IsEnabled => isEnabled;
+
+		public RulesetWatcher(World world, ModData modData)
+		{
+			this.world = world;
+			this.modData = modData;
+
+			var files = modData.Manifest.Rules
+				.Concat(modData.Manifest.Weapons)
+				.Concat(modData.Manifest.Sequences);
+			foreach (var file in files)
+			{
+				string filename;
+				using (var fs = modData.DefaultFileSystem.Open(file) as FileStream)
+					filename = fs?.Name;
+
+				if (string.IsNullOrEmpty(filename))
+					continue;
+
+				var fullPath = Path.GetFullPath(filename);
+				watchFiles[fullPath] = file;
+			}
+
+			timer = new Timer(10) { AutoReset = false };
+			timer.Elapsed += RefreshTimer;
+
+			watcher = new FileSystemWatcher(modData.Manifest.Package.Name)
+			{
+				IncludeSubdirectories = true
+			};
+			watcher.Changed += FileChanged;
+
+			foreach (var file in watchFiles.Keys.Distinct())
+			{
+				watcher.Filters.Add(Path.GetFileName(file));
+			}
+		}
+
+		public void StartWatching()
+		{
+			if (isDisposed)
+				throw new ObjectDisposedException(nameof(RulesetWatcher));
+
+			watcher.EnableRaisingEvents = true;
+			isEnabled = true;
+			RestartTimerIfEnabled();
+		}
+
+		public void StopWatching()
+		{
+			if (isDisposed)
+				throw new ObjectDisposedException(nameof(RulesetWatcher));
+
+			isEnabled = false;
+			timer.Stop();
+			watcher.EnableRaisingEvents = false;
+
+			lock (fileQueue)
+				fileQueue.Clear();
+		}
+
+		void RefreshTimer(object sender, ElapsedEventArgs e)
+		{
+			if (fileQueue.Count == 0)
+			{
+				RestartTimerIfEnabled();
+				return;
+			}
+
+			List<string> copy;
+			lock (fileQueue)
+			{
+				if (fileQueue.Count == 0)
+					return;
+
+				copy = fileQueue.ToList();
+				fileQueue.Clear();
+			}
+
+			if (isDisposed || !isEnabled)
+				return;
+
+			Game.RunAfterTick(() =>
+			{
+				if (isDisposed || !isEnabled)
+					return;
+
+				var modFsFilenames = copy.Select(f => watchFiles[f]).ToArray();
+
+				var defaultRules = world.Map.Rules;
+				var rulesFiles = FindRulesetFiles(modData.Manifest.Rules, modFsFilenames).ToArray();
+				var weaponFiles = FindRulesetFiles(modData.Manifest.Weapons, modFsFilenames).ToArray();
+				var sequenceFile = FindRulesetFiles(modData.Manifest.Sequences, modFsFilenames).FirstOrDefault();
+
+				if (rulesFiles.Length > 0)
+					defaultRules.LoadActorTraitsFromRuleFile(world, modData, rulesFiles);
+				else if (weaponFiles.Length > 0)
+					defaultRules.LoadWeaponsFromFile(world, modData, rulesFiles);
+				else if (sequenceFile != null)
+					world.Map.Sequences.ReloadSequenceSetFromFiles(modData.DefaultFileSystem, sequenceFile);
+
+				RestartTimerIfEnabled();
```

Comment:
I think this would be cleaner if it had the `if` statement in-line and the method was simply `RestartTimer()`, because the performance is the same yet having `RestartTimer()` on its own means we can re-use it for other purposes if we want.

Reply:
Not sure I follow. The timer needs to enabled *only and only if* the `RulesetWatcher` itself is enabled. We don't want to restart the timer, if `RulesetWatcher` has been disabled in the mean time (i.e. between `if (isDisposed || !isEnabled)` above and the end of `RunAfterTick()` action that does the reload).

Reply:
I don't mean change the logic, I just mean take the `If` out of the timer restart and put it outside, then rename the method to simply `RestartTimer() `

Reply:
`RestartTimerIfEnabled()` is used from three places (two of which need to take into account `isEnabled` state). I don't see any benefit to having one-line method `RestartTimer()`, which just calls `timer.Start()`.

In any case, this is the code that would benefit greatly from code review done by multithreading expert (like @RoosterDragon).

Reply:
I explained the benefits. It's also unconventional to write methods with specific conditions in their name because they are not very reusable. It just so happens your code does this 3 times, but other code may not. Also if you are already checking for whether something is enabled then you are doubling up on the check.

Reply:
The timer has been removed, therefore the discussion is now mute.